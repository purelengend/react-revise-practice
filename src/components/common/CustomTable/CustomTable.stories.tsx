import type { Meta, StoryObj } from "@storybook/react";

// Components
import CustomTable from ".";

const meta: Meta<typeof CustomTable> = {
  title: "Components/CustomTable",
  component: CustomTable,
  argTypes: {
    columns: {
      description: "Array of column to render in the table head",
    },
    data: {
      description: "Array of data to render in the table cells",
    },
    isFetching: {
      description: "Whether the table is fetching data",
    },
  },
  args: {
    columns: [
      {
        title: "Name",
        key: "name",
      },
      {
        title: "Email",
        key: "email",
      },
      {
        title: "Phone",
        key: "phone",
      },
    ],
    data: [
      {
        name: "example",
        email: "example@gmail.com",
        phone: "0909090909",
      },
    ],
    isFetching: false,
  },
  parameters: {
    controls: {
      expanded: true,
    },
  },
};

export default meta;

type Story = StoryObj<typeof CustomTable>;

export const Primary: Story = {};
