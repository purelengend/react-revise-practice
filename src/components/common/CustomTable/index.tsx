import {
  Table,
  TableContainer,
  Tbody,
  Td,
  Th,
  Thead,
  Tr,
} from "@chakra-ui/react";
import { ReactElement, useMemo } from "react";

export type ColumnProps<T> = {
  key: string;
  title: string | ReactElement;
  render?: (item: T) => ReactElement;
};

export type TableProps<T> = {
  columns: Array<ColumnProps<T>>;
  data?: Array<T>;
  isFetching: boolean;
};

const CustomTable = <T,>({ columns, data, isFetching }: TableProps<T>) => {
  const headers = useMemo(
    () =>
      columns.map((column) => {
        return (
          <Th minW={30} key={`tableHeadCell-${column.title}`} color="gray.200">
            {column.title}
          </Th>
        );
      }),
    [columns],
  );

  const rows = useMemo(
    () =>
      !data?.length ? (
        <Tr opacity={isFetching ? 0.5 : 1}>
          <Td colSpan={columns.length} textAlign="center">
            No data
          </Td>
        </Tr>
      ) : (
        data.map((row, rowIndex) => {
          return (
            <Tr
              opacity={isFetching ? 0.5 : 1}
              bg="white"
              h={21.25}
              borderRadius="lg"
              key={`row-${rowIndex}`}
              sx={{
                "& td:first-of-type, & td:last-of-type": {
                  borderRadius: "lg",
                },
              }}
            >
              {columns.map((column, columnIndex) => {
                const value = column.render
                  ? column.render(row as T)
                  : (row[column.key as keyof typeof row] as string);

                return <Td key={`cell-${columnIndex}`}>{value}</Td>;
              })}
            </Tr>
          );
        })
      ),
    [columns, data, isFetching],
  );
  return (
    <TableContainer>
      <Table
        w="full"
        __css={{
          borderCollapse: "separate",
          borderSpacing: "0 10px",
        }}
        variant="unstyled"
      >
        <Thead h={21.25}>
          <Tr>{headers}</Tr>
        </Thead>
        <Tbody>{rows}</Tbody>
      </Table>
    </TableContainer>
  );
};

export default CustomTable;
