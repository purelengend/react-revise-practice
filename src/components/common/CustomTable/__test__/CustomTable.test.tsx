import { render } from "@testing-library/react";

// Component
import CustomTable from "..";

describe("CustomTable test cases", () => {
  const mockColumn = [
    {
      title: "Mock Title",
      key: "value",
    },
    {
      title: "Mock render title",
      key: "mock",
      render: () => <></>,
    },
  ];

  const mockData: Array<{
    value: string;
    mock: string;
  }> = [
    {
      value: "mock",
      mock: "mock",
    },
  ];

  const setup = (
    data: Array<{
      value: string;
      mock: string;
    }>,
    isFetching: boolean,
  ) =>
    render(
      <CustomTable columns={mockColumn} data={data} isFetching={isFetching} />,
    );

  it("should render correctly", () => {
    const { container } = setup(mockData, false);

    expect(container).toMatchSnapshot();
  });

  it("should be loading when fetching data", () => {
    const { container } = setup(mockData, true);

    expect(container).toMatchSnapshot();
  });

  it("should handle loading state when the page is first loaded", () => {
    const { container } = setup([], true);

    expect(container).toMatchSnapshot();
  });

  it("should handle message when no data is fetched", () => {
    const { container } = setup([], false);

    expect(container).toMatchSnapshot();
  });
});
