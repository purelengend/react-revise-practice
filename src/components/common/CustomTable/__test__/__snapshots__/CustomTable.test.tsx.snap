// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomTable test cases should be loading when fetching data 1`] = `
<div>
  <div
    class="chakra-table__container css-zipzvv"
  >
    <table
      class="chakra-table css-1xlmhul"
    >
      <thead
        class="css-1mwmt6s"
      >
        <tr
          class="css-0"
        >
          <th
            class="css-gflm3x"
          >
            Mock Title
          </th>
          <th
            class="css-gflm3x"
          >
            Mock render title
          </th>
        </tr>
      </thead>
      <tbody
        class="css-0"
      >
        <tr
          class="css-17vy9fv"
        >
          <td
            class="css-0"
          >
            mock
          </td>
          <td
            class="css-0"
          />
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`CustomTable test cases should handle loading state when the page is first loaded 1`] = `
<div>
  <div
    class="chakra-table__container css-zipzvv"
  >
    <table
      class="chakra-table css-1xlmhul"
    >
      <thead
        class="css-1mwmt6s"
      >
        <tr
          class="css-0"
        >
          <th
            class="css-gflm3x"
          >
            Mock Title
          </th>
          <th
            class="css-gflm3x"
          >
            Mock render title
          </th>
        </tr>
      </thead>
      <tbody
        class="css-0"
      >
        <tr
          class="css-1tzeee1"
        >
          <td
            class="css-xi606m"
            colspan="2"
          >
            No data
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`CustomTable test cases should handle message when no data is fetched 1`] = `
<div>
  <div
    class="chakra-table__container css-zipzvv"
  >
    <table
      class="chakra-table css-1xlmhul"
    >
      <thead
        class="css-1mwmt6s"
      >
        <tr
          class="css-0"
        >
          <th
            class="css-gflm3x"
          >
            Mock Title
          </th>
          <th
            class="css-gflm3x"
          >
            Mock render title
          </th>
        </tr>
      </thead>
      <tbody
        class="css-0"
      >
        <tr
          class="css-whh5e5"
        >
          <td
            class="css-xi606m"
            colspan="2"
          >
            No data
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`CustomTable test cases should render correctly 1`] = `
<div>
  <div
    class="chakra-table__container css-zipzvv"
  >
    <table
      class="chakra-table css-1xlmhul"
    >
      <thead
        class="css-1mwmt6s"
      >
        <tr
          class="css-0"
        >
          <th
            class="css-gflm3x"
          >
            Mock Title
          </th>
          <th
            class="css-gflm3x"
          >
            Mock render title
          </th>
        </tr>
      </thead>
      <tbody
        class="css-0"
      >
        <tr
          class="css-p2advv"
        >
          <td
            class="css-0"
          >
            mock
          </td>
          <td
            class="css-0"
          />
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
